packages/go/pocket_rpc/client_pool.go

```diff
@@ -0,0 +1,170 @@
+package pocket_rpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/rs/zerolog"
+	"golang.org/x/time/rate"
+	"math"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+type Backoff struct {
+	min time.Duration
+	max time.Duration
+}
+
+func NewBackoff(min, max time.Duration) *Backoff {
+	return &Backoff{min: min, max: max}
+}
+
+func (b *Backoff) Duration(n int) time.Duration {
+	if n == 0 {
+		return b.min
+	}
+	duration := time.Duration(math.Pow(2, float64(n))) * b.min
+	if duration > b.max {
+		return b.max
+	}
+	return duration
+}
+
+type Client struct {
+	url    string
+	client *http.Client
+}
+
+func (c Client) Do(req *http.Request) (*http.Response, error) {
+	return c.client.Do(req)
+}
+
+type ClientPoolOptions struct {
+	MaxRetries int
+	ReqPerSec  int
+	MinBackoff time.Duration
+	MaxBackoff time.Duration
+}
+
+type ClientPool struct {
+	clients chan *Client
+	logger  *zerolog.Logger
+	opts    *ClientPoolOptions
+}
+
+func NewDefaultClientPoolOptions() *ClientPoolOptions {
+	return &ClientPoolOptions{
+		MaxRetries: 3,
+		ReqPerSec:  10,
+		MinBackoff: 10 * time.Second,
+		MaxBackoff: 60 * time.Second,
+	}
+}
+
+func NewClientPool(servers []string, opts *ClientPoolOptions, logger *zerolog.Logger) *ClientPool {
+	if opts == nil {
+		opts = NewDefaultClientPoolOptions()
+	}
+
+	cp := ClientPool{
+		clients: make(chan *Client, len(servers)),
+		logger:  logger,
+		opts:    opts,
+	}
+
+	for _, server := range servers {
+		cp.setClient(&Client{url: server, client: &http.Client{Timeout: 1 * time.Minute}})
+	}
+
+	return &cp
+}
+
+func (cp *ClientPool) DoRRLoadBalanced(req *http.Request, timeout time.Duration) (*http.Response, int, error) {
+	retries := 0
+	backoff := NewBackoff(cp.opts.MinBackoff, cp.opts.MaxBackoff) // adjust min and max duration as necessary
+
+	for i := 0; i < len(cp.clients); i++ {
```

Comment:
This cycle is over available clients, but before trying a new client, the backoff sleep is triggered. 
This means that if client A fails, then we will wait some time before trying with client B. Is this intended?
If there are many clients, some of them will not be used, for example if `len(cp.clients)` is greater than `cp.opts.MaxRetries`

Reply:
I will make the backoff only be applied if the for complete at least 1 round